### PR TITLE
Converted bunyan calls to JSON format !:

### PR DESCRIPTION
**Pull request by garypgryan**

### CODE REVIEW

File: PrettyPrintDuplexStreamJson.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+// make a class that implements a Duplex stream; this means you can use it to pipe into and out of
+class PrettyPrintDuplexStreamJson extends Transform {
+  constructor(name, options) {
+    super(options);
+    this.name = name;
+    this.solutionMap = {};
+    this.ruleMap = {};
+    this.templateStrings = {};
+    const csvFilePath = path.join(__dirname, 'errorMessages.txt' );
+    this.eCode = '-1';   // put in a value for error messages that don't have an ERROR_CODE
```
Comment (cmoesel) [inline PrettyPrintDuplexStreamJson.js]: I don't understand why we have `eCode` as a member of `PrettyPrintDuplexStreamJson`.  I understand why you register it in the `solutionMap` and `idSet` -- but I don't think it needs to be a member of the class?

File: PrettyPrintDuplexStreamJson.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      let parts= line.split(',');
+      let key = parts[keyColumnNumber].trim();
+      let value = parts[valueColumnNumber];
+      map[key] = value;
+    }
+  }
+
+  translateNames( inName ) {   // translate error Strings to friendlier informative alternatives.
+    if (inName === 'shr-expand') {
+      return('FHIR Mapping(expansion)');
```
Comment (cmoesel) [inline PrettyPrintDuplexStreamJson.js]: The name for `shr-expand` should be `Model Expansion`

File: PrettyPrintDuplexStreamJson.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      }
+      else {
+        myECode = result[1].trim();   
+      }
+    }
+    return(myECode);
+  }
+
+  getAttributeOrEmptyString( myPart) {   // guard against undefined or null attributes
+    if (myPart === undefined || myPart === null) {
```
Comment (cmoesel) [inline PrettyPrintDuplexStreamJson.js]: If you just say `myPart == null` (note, `==` not `===`) then that covers both these cases (`null` and `undefined`).

File: PrettyPrintDuplexStreamJson.js [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    else {
+      return(myPart);
+    }
+  }
+
+  buildHashKey( errorCode, ruleMap,  jsonObj ) {  // construct a key by concatenating fields; this key defines 'uniqueness' for deduplication
+    let hashValue = '';
+    let semicolonDelimitedKeyList = ruleMap[errorCode];
+
+    if (semicolonDelimitedKeyList !== undefined && semicolonDelimitedKeyList !== null ) {
```
Comment (cmoesel) [inline PrettyPrintDuplexStreamJson.js]: I won't mention it again after this, but instead of checking against `null` and `undefined` separately, you can usually use the looser versions of `==` and `!=` against `null` (which will also cover `undefined`).

File: PrettyPrintDuplexStreamJson.js [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    this.oldErrCodeRegex = /ERROR_CODE:([\d]+)\s*/; //match ERROR_CODE:ddddd  
+    this.myJson = JSON.parse(myinline);          // convert String to object, then grab attributes
+    let jsonKeys = Object.keys(this.myJson);
+    this.jMsg = this.getAttributeOrEmptyString(this.myJson.msg ); 
+    this.modulePart = this.getAttributeOrEmptyString(this.myJson.module);  //grab module
+    this.errCodeRegex = /([\d]{5})\s*/; //match ERROR_CODE:ddddd  exactly 5 digits!!!!
+    this.eCode = this.parseErrorCode(this.errCodeRegex, this.jMsg);   //extract ERROR_CODE:ddddd from msg attribute
+    this.detailMsg = this.jMsg.replace(this.oldErrCodeRegex,'').replace(this.errCodeRegex, '');        // remove ERROR_CODE:ddddd from msg to get message detail
+    this.myJsonErrorCode = this.getAttributeOrEmptyString(this.myJson.$ERROR_CODE ).toString(); 
+    this.myTemplate = this.templateStrings[this.eCode];
```
Comment (cmoesel) [inline PrettyPrintDuplexStreamJson.js]: I'm confused.  Why are all of these things stored on `this`?  Do they really need to be stored on the class object itself?  Can't they just be local variables?

I also find the code above a little hard to follow.  It seems like you're parsing the error code in multiple ways in multiple places.  Maybe it's right (I'm not sure), but some more documentation might be helpful.

**Comment by cmoesel** [comment]
It looks like my comments from Feb 18 and Feb 21 haven't been addressed yet.  I do plan to review this again, but you can start by addressing the comments above.

Also, if you're comfortable with it, I'd suggest rebasing on the latest master.

File: PrettyPrintDuplexStreamJson.js (hunk 1 above)
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: I removed most of the variables from the class object.

File: PrettyPrintDuplexStreamJson.js (hunk 2 above)
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: Changed name to 'Model Expansion'

File: PrettyPrintDuplexStreamJson.js (hunk 3 above)
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: I changed all of those instances to use the more concise form.

File: PrettyPrintDuplexStreamJson.js (hunk 4 above)
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: I changed all of the instances to the concise form.

File: PrettyPrintDuplexStreamJson.js (hunk 5 above)
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: I removed some of the obsolete code and changed the variables to local where applicable.

File: PrettyPrintDuplexStream.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
+const Transform = require('stream').Transform;
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply PrettyPrintDuplexStream.js]: After reviewing further, I see you use this depending on the logging mode.  Do you think that's necessary?  I'm just wondering whether it's worth supporting it in both -- as there's some duplicate code and it's more to maintain.

Honest question -- looking for your opinion.

File: PrettyPrintDuplexStreamJson.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    this.idSet = new Set();
+    this.idSet.add( noErrorCode );
+  }
+
+  buildMapFromFile( filePath, map, keyColumnNumber, valueColumnNumber) {
+    const recArray = fs.readFileSync(filePath).toString().split('\n'); 
+    // populate a map with the key as the ERROR_CODE number and the value is the suggested solution
+    for( const i in recArray) {   
+      const line = recArray[i].toString();
+      const parts= line.split(',');
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: I agree.   That's something I want to address, as well as support comments that start with
"//" in the error message file.

File: PrettyPrintDuplexStream.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStream.js]: I left the regex version in place as a transitional stage until we get all the calls to logger.error upgraded to the new JSON API.   I don't think there is value in the long term keeping it around once we' ve switched to JSON.

File: PrettyPrintDuplexStreamJson.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    const eCode = this.parseErrorCode( /([\d]{5})\s*/, jMsg); //extract ERROR_CODE:ddddd from msg attribute;eCode = noErrorCode if not found;
+    let detailMsg = '';
+    let myTemplate = this.templateStrings[ eCode ];
+
+    if (myTemplate != null) {
+      myTemplate = this.processTemplate_new(jsonKeys, myTemplate, myJson  ) ;
+      detailMsg = myTemplate;
+    }
+    else {
+      console.log( errorCodeColor(' Message is missing errorCode; no template found.  Default error code '+ eCode + ':'));
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: No, that was not intentional; I will fix it.

File: PrettyPrintDuplexStreamJson.js [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    // populate a map with the key as the ERROR_CODE number and the value is the suggested solution
+    this.idSet = new Set();
+    this.idSet.add( noErrorCode );
+  }
+
+  buildMapFromFile( filePath, map, keyColumnNumber, valueColumnNumber) {
+    const recArray = fs.readFileSync(filePath).toString().split('\n'); 
+    // populate a map with the key as the ERROR_CODE number and the value is the suggested solution
+    for( const i in recArray) {   
+      const line = recArray[i].toString();
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: When I remove .toString(); the next line throws an error saying there is no split function; I guess it's doing a type conversion to String that is needed.

const line = recArray[i].toString();
      const parts= line.split(',');

[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: TypeError: line.split is not a function
    at PrettyPrintDuplexStreamJson.buildMapFromFile (/Users/ggryan/Downloads/prettyPrintJson/shr-cli/PrettyPrintDuplexStreamJson.js:38:25)
    at new PrettyPrintDuplexStreamJson (/Users/ggryan/Downloads/prettyPrintJson/shr-cli/PrettyPrintDuplexStreamJson.js:24:10)
    at Object.<anonymous> (/Users/ggryan/Downloads/prettyPrintJson/shr-cli/app.js:77:19)
    at Module._compile (internal/modules/cjs/loader.js:722:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:733:10)
    at Module.load (internal/modules/cjs/loader.js:620:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:560:12)
    at Function.Module._load (internal/modules/cjs/loader.js:552:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:775:12)
    at startup (internal/bootstrap/node.js:300:19)

File: PrettyPrintDuplexStreamJson.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    this.solutionMap = {};
+    this.ruleMap = {};
+    this.templateStrings = {};
+    const csvFilePath = path.join(__dirname, 'errorMessages.txt' );
+    this.solutionMap[ noErrorCode ] = 'Error message has no error code; please add error code';  // -1 means noErrorCode
+    // build the hashMap resources from the errorMessages csv file; each column has a part 
+    this.buildMapFromFile( csvFilePath, this.solutionMap, 0, 2);
+    this.buildMapFromFile( csvFilePath, this.ruleMap, 0, 3);
+    this.buildMapFromFile( csvFilePath, this.templateStrings, 0, 1);
+    // populate a map with the key as the ERROR_CODE number and the value is the suggested solution
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: Corrected comment to use "set", not map.    It had been a map earlier, but the comment was never updated.

File: PrettyPrintDuplexStreamJson.js [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+      const value = parts[valueColumnNumber];
+      map[key] = value;
+    }
+  }
+
+  translateNames( inName ) {   // translate error Strings to friendlier informative alternatives.
+    switch(inName) {
+    case 'shr-expand': return ('Model Expansion');
+    case 'shr-fhir-export' : return ('FHIR Export') ;
+    default: return(inName);
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: Fixed all return statements and removed unnecessary parentheses.

File: PrettyPrintDuplexStreamJson.js [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    const jsonKeys = Object.keys( myJson);
+    const jMsg = this.getAttributeOrEmptyString( myJson.msg ); 
+    const modulePart = this.getAttributeOrEmptyString( myJson.module);  //grab module
+    const eCode = this.parseErrorCode( /([\d]{5})\s*/, jMsg); //extract ERROR_CODE:ddddd from msg attribute;eCode = noErrorCode if not found;
+    let detailMsg = '';
+    let myTemplate = this.templateStrings[ eCode ];
+
+    if (myTemplate != null) {
+      myTemplate = this.processTemplate_new(jsonKeys, myTemplate, myJson  ) ;
+      detailMsg = myTemplate;
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: renamed function to processTemplate.

File: PrettyPrintDuplexStreamJson.js [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+            console.log('undefined JSON atttribute ' + parts[i].trim());
+          }
+          hashValue += curKey + '$';    
+        }
+      } 
+    }
+    return(hashValue);
+  }
+
+  processTemplate_new(jsonKeys, myTemplate, myJson  ) {  // get the elements from the template and fill them in; return the detail message
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: renamed function to processTemplate.

File: PrettyPrintDuplexStreamJson.js [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        }
+      }
+    } 
+    return(template);
+  }
+  
+  // this function processes a single error message and returns a colorized, formatted string
+  // for the moment, it writes the original input message in red to console.log
+  processLine(myinline, printAllErrors) {
+    const myline = myinline.toString();
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: no, I removed it.

File: PrettyPrintDuplexStreamJson.js [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+      return(outline);
+    }
+    else if (this.idSet.has(myDedupHashKey)) {
+      return('');
+    }
+    else {
+      this.idSet.add(myDedupHashKey);
+      if (printAllErrors !== false) {
+        console.log( originalErrorColor (myline )); 
+      }
```
[the remark this answers is not in the document]
Comment (garypgryan) [reply PrettyPrintDuplexStreamJson.js]: I moved the "printAllErrors" logic to the top of the method, so exiting the method will still print the original when "printAllErrors" is true.

**Comment by ngfreiter** [comment]
This PR has been replaced by #227.
